runtime/commonMain/src/kotlinx/serialization/SerialDescriptor.kt

```diff
@@ -0,0 +1,201 @@
+     *
+     * For generated serializers, serial name is equal to the corresponding class's fully-qualified name
+     * or, if overridden [SerialName].
+     * Custom serializers should provide a unique serial name that identify both the serializable class and
+     * the serializer itself.
+     */
+    public val serialName: String
+
+    @Deprecated(
+        message = "name property deprecated in the favour of serialName",
+        level = DeprecationLevel.ERROR,
+        replaceWith = ReplaceWith("serialName")
+    )
+    public val name: String
+        get() = serialName
+
+    /**
+     * The kind of the serialized form that determines **the shape** of the serialized data.
+     * Formats use serial kind to add and parse serializer-agnostic metadata to the result.
+     *
+     * For example, JSON format wraps [classes][StructureKind.CLASS] and [StructureKind.MAP] into
+     * brackets, while ProtoBuf just serialize these types in a separate ways.
+     *
+     * Kind should be consistent with the implementation, for example, if it is a [primitive][PrimitiveKind],
+     * then its elements count should be zero and vice versa.
+     */
+    public val kind: SerialKind
+
+    /**
+     * Whether the descriptor describes nullable element.
+     * Returns `true` if associated serializer can serialize/deserialize nullable elements of the described type.
+     */
+    public val isNullable: Boolean get() = false
+
+    /**
+     * The number of elements this descriptor describes, besides from the class itself.
+     * [elementsCount] describes the number of **semantic** elements, not the number
+     * of actual fields/properties in the serialized form, even though they frequently match.
+     *
+     * For example, for the following class
+     * `class Complex(val real: Long, val imaginary: Long)` the corresponding descriptor
+     * and the serialized form both have two elements, while for `class IntList : ArrayList<Int>()`
+     * the corresponding descriptor has a single element (`IntDescriptor`, the type of list element),
+     * but from zero up to `Int.MAX_VALUE` values in the serialized form.
+     */
+    public val elementsCount: Int
+
+    /**
+     * Returns a serial annotations of the associated class.
+     * Serial annotations can be used to specify an additional
+     * metadata that may be used during serialization, for example a [serial id][SerialId].
+     */
+    public val annotations: List<Annotation> get() = emptyList()
+
+    /**
+     * Returns a _positional_ name of the child at the given [index].
+     * Positional name usually represents a corresponding property name in the class, associated with
+     * the current descriptor.
+     *
+     * This method may be inconsistent with [elementsCount] for collection and map types. Refer to its documentation
+     * for more details.
+     *
+     * @throws IndexOutOfBoundsException for an illegal [index] values.
```

Comment:
Should the documentation state that valid values are `0 until elementsCount` and that both should maintain that invariant? I know that older implementations did violate it.